Lib/tarfile.py

```diff
@@ -1943,7 +1945,7 @@ def add(self, name, arcname=None, recursive=True, *, filter=None):
             arcname = name
 
         # Skip if somebody tries to archive the archive...
-        if self.name is not None and os.path.abspath(name) == self.name:
+        if self.name is not None and os.path.abspath(os.fspath(name)) == self.name:
```

Comment:
Please move `os.fspath(name)` to the very beginning of the function, just after `self._check(...)`.